File: server/storage.py
"""Bounded read caches and validated, portable book snapshots."""
from __future__ import annotations

import base64
import hashlib
import json
import os
import re
import tempfile
import threading
from collections import OrderedDict
from pathlib import Path

ASSET_LIMIT = 32 * 1024 * 1024
ASSETS_LIMIT = 96 * 1024 * 1024
MAX_RECORDS = 1_000_000
ASSET_NAME = re.compile(r'[A-Za-z0-9_-][A-Za-z0-9_.-]{0,127}\Z')


def signature(path: Path):
    st = path.stat()
    return (st.st_mtime_ns, st.st_size, st.st_ino)


class JsonCache:
    """Budget in serialized bytes; oversized bodies are deliberately not retained."""
    def __init__(self, count=128, budget=None, item_limit=None):
        self.count = count
        self.budget = budget if budget is not None else int(os.environ.get('JSON_CACHE_BYTES', 96 * 1024 * 1024))
        self.item_limit = item_limit if item_limit is not None else int(os.environ.get('JSON_CACHE_ITEM_BYTES', 64 * 1024 * 1024))
        self.entries = OrderedDict()
        self.size = 0
        self.lock = threading.RLock()
        self.loads = [threading.RLock() for _ in range(8)]

    def get(self, path, _attempt=0):
        path = Path(path)
        try:
            stamp = signature(path)
        except FileNotFoundError:
            self.evict(path)
            return None
        key = str(path)
        with self.lock:
            hit = self.entries.get(key)
            if hit and hit[0] == stamp:
                self.entries.move_to_end(key)
                return hit[1]
        # Coalesce concurrent reads of one large book; do not multiply parse allocations.
        with self.loads[hash(key) % len(self.loads)]:
            return self._load(path, key, stamp, _attempt)

    def _load(self, path, key, stamp, _attempt):
        with self.lock:
            hit = self.entries.get(key)
            if hit and hit[0] == stamp:
                self.entries.move_to_end(key)
                return hit[1]
        value = json.loads(path.read_text(encoding='utf-8'))
        if signature(path) != stamp:
            if _attempt >= 2:
                raise ValueError('文件正在更新，请稍后重试')
            return self.get(path, _attempt + 1)
        if stamp[1] <= self.item_limit:
            with self.lock:
                self.evict(path)
                self.entries[key] = (stamp, value)
                self.size += stamp[1]
                while len(self.entries) > self.count or self.size > self.budget:
                    _, (old, _) = self.entries.popitem(last=False)
                    self.size -= old[1]
        return value

    def evict(self, path):
        key = str(path)
        with self.lock:
            for k in list(self.entries):
                if k == key or k.startswith(key + os.sep):
                    stamp, _ = self.entries.pop(k)
                    self.size -= stamp[1]


def write_json(path: Path, value):
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp = tempfile.mkstemp(prefix='.' + path.name + '.', suffix='.tmp', dir=path.parent)
    try:
        with os.fdopen(fd, 'w', encoding='utf-8') as f:
            json.dump(value, f, ensure_ascii=False, separators=(',', ':'))
            f.flush()
            os.fsync(f.fileno())
        os.replace(tmp, path)
    finally:
        if os.path.exists(tmp):
            os.unlink(tmp)


_BYLINE = re.compile(r'^(?P<title>.*?)\s*[-_—\s]*作者\s*[:：]\s*(?P<author>.*?)\s*$')
_SITE = re.compile(r'(?i)^[\w.-]+\.(com|net|org|cn|cc)$')


def display_title(title, author):
    """TXT file names often carry the byline ("书名 作者：某某"); show it as title + author.
    Display only: stored book data (and the ids derived from it) stay unchanged."""
    title, author = (title or '').strip(), (author or '').strip()
    m = _BYLINE.match(title)
    if m and m.group('title'):
        title = m.group('title')
        found = m.group('author').strip()
        if not author and found and not _SITE.match(found):
            author = found
    return title, author


def shelf_fields(book):
    chapters = book.get('chapters') or []
    out = {k: book.get(k) for k in ('title', 'author', 'len', 'cover', 'lang', 'genre')}
    if not out['lang']:
        from pipeline.lang import book_lang
        out['lang'] = book_lang(book)
    out.update(chapters=len(chapters), thin=(book.get('len') or 0) < 30000 or
               sum(c.get('kind') == 'body' for c in chapters) < 2)
    return out


def shelf_metadata(root, cache):
    # Cold shelf requests should not simultaneously parse the same large legacy body.
    with cache.loads[hash(str(root / 'book.json')) % len(cache.loads)]:
        return _shelf_metadata(root, cache)


def _shelf_metadata(root, cache):
    source = root / 'book.json'
    stamp = list(signature(source))
    target = root / 'shelf.json'
    stored = cache.get(target)
    if stored and stored.get('source') == stamp and stored.get('book', {}).get('lang'):
        return stored['book']
    # One sequential rebuild for legacy books; large text is not put in the read cache.
    book = json.loads(source.read_text(encoding='utf-8'))
    if list(signature(source)) != stamp:
        return shelf_metadata(root, cache)
    fields = shelf_fields(book)
    write_json(target, {'source': stamp, 'book': fields})
    return fields


def integer(value, label, low=0, high=120_000_000):
    if isinstance(value, bool) or not isinstance(value, int) or not low <= value <= high:
        raise ValueError(f'{label}无效')
    return value


def asset_name(name):
    if not isinstance(name, str) or not ASSET_NAME.fullmatch(name):
        raise ValueError('图片文件名无效')
    return name


def validate_book(book):
    if not isinstance(book, dict) or not isinstance(book.get('title'), str):
        raise ValueError('书籍信息无效')
    length = integer(book.get('len'), '正文长度')
    blocks, chapters = book.get('blocks'), book.get('chapters')
    if not isinstance(blocks, list) or not blocks or len(blocks) > MAX_RECORDS:
        raise ValueError('正文段落无效')
    if not isinstance(chapters, list) or not chapters or len(chapters) > 100_000:
        raise ValueError('章节无效')
    previous = -1
    for block in blocks:
        if not isinstance(block, dict) or block.get('k') not in ('p', 'h', 'img') or not isinstance(block.get('t'), str):
            raise ValueError('正文段落格式无效')
        pos = integer(block.get('o'), '段落位置', 0, length)
        if pos < previous or pos + len(block['t'].encode('utf-16-le')) // 2 > length:
            raise ValueError('段落位置超出正文')
        previous = pos
        if block['k'] == 'img':
            asset_name(block.get('src'))
        for item in block.get('fn', []):
            if not isinstance(item, list) or len(item) != 2 or not isinstance(item[1], str):
                raise ValueError('脚注索引无效')
            integer(item[0], '脚注位置', 0, len(block['t'].encode('utf-16-le')) // 2)
    previous = -1
    for chapter in chapters:
        if not isinstance(chapter, dict) or not isinstance(chapter.get('title'), str):
            raise ValueError('章节格式无效')
        a = integer(chapter.get('b0'), '章节起点', 0, len(blocks))
        b = integer(chapter.get('b1'), '章节终点', a, len(blocks))
        p = integer(chapter.get('o0'), '章节位置', 0, length)
        integer(chapter.get('o1'), '章节终点', p, length)
        if a == b or p < previous:
            raise ValueError('章节顺序无效')
        previous = p
    if not isinstance(book.get('notes', {}), dict):
        raise ValueError('脚注格式无效')
    if not all(isinstance(k, str) and isinstance(v, str) for k, v in book.get('notes', {}).items()):
        raise ValueError('脚注内容无效')
    if book.get('cover'):
        asset_name(book['cover'])
    return book


def validate_graph(kg, length):
    if not isinstance(kg, dict) or not isinstance(kg.get('log', []), list):
        raise ValueError('人物图谱格式无效')
    log = kg.get('log', [])
    if len(log) > MAX_RECORDS:
        raise ValueError('人物图谱太大')
    fields = {'person': ('id', 'name'), 'name': ('id', 'name'), 'alias': ('id', 'alias'),
              'merge': ('from', 'into'), 'profile': ('id',), 'attr': ('id', 'key', 'value'),
              'rel': ('a', 'b'), 'event': ('text',), 'imp': ('id',), 'cnt': (),
              'recap': ('text',), 'saga': ('text',)}
    previous = -1
    for row in log:
        if not isinstance(row, dict) or row.get('t') not in fields:
            raise ValueError('人物记录类型无效')
        pos = integer(row.get('p'), '人物记录位置', 0, length)
        if pos < previous:
            raise ValueError('人物记录没有按位置排序')
        previous = pos
        for key in fields[row['t']]:
            if not isinstance(row.get(key), str):
                raise ValueError('人物记录字段无效')
        for key in ('intro', 'tagline', 'bio', 'family', 'a_is', 'b_is', 'desc', 'status', 'reason'):
            if key in row and row[key] is not None and not isinstance(row[key], str):
                raise ValueError('人物记录文本无效')
        if 's' in row:
            integer(row['s'], '原文位置', 0, length)
        if 'imp' in row:
            integer(row['imp'], '人物重要度', 0, 100)
        if 'chapter' in row:
            integer(row['chapter'], '章节编号', 0, 100_000)
        if row['t'] == 'event' and (not isinstance(row.get('who'), list) or
                                   not all(isinstance(x, str) for x in row['who'])):
            raise ValueError('事件人物无效')
        if row['t'] == 'cnt' and (not isinstance(row.get('c'), dict) or
                                 not all(isinstance(k, str) and isinstance(v, int) and v >= 0 for k, v in row['c'].items())):
            raise ValueError('人物计数无效')
    return kg


def referenced_assets(book):
    names = {b['src'] for b in book['blocks'] if b['k'] == 'img'}
    if book.get('cover'):
        names.add(book['cover'])
    return names


def encode_assets(root, book):
    assets, total = {}, 0
    for name in sorted(referenced_assets(book)):
        asset_name(name)
        path = root / 'img' / name
        if path.is_symlink() or not path.is_file():
            raise ValueError('这本书缺少图片，请先修复后再导出')
        size = path.stat().st_size
        total += size
        if size > ASSET_LIMIT or total > ASSETS_LIMIT:
            raise ValueError('图片超过导出限制')
        raw = path.read_bytes()
        assets[name] = {'size': len(raw), 'sha256': hashlib.sha256(raw).hexdigest(),
                        'base64': base64.b64encode(raw).decode('ascii')}
    return assets


def decode_assets(assets, book):
    if not isinstance(assets, dict) or set(assets) != referenced_assets(book):
        raise ValueError('图片清单与正文不一致')
    out, total = {}, 0
    for name, item in assets.items():
        asset_name(name)
        if not isinstance(item, dict):
            raise ValueError('图片格式无效')
        size = integer(item.get('size'), '图片长度', 0, ASSET_LIMIT)
        total += size
        if total > ASSETS_LIMIT:
            raise ValueError('图片总量太大')
        encoded = item.get('base64')
        if not isinstance(encoded, str) or len(encoded) > ((size + 2) // 3) * 4:
            raise ValueError('图片编码长度无效')
        try:
            raw = base64.b64decode(encoded, validate=True)
        except Exception as exc:
            raise ValueError('图片编码无效') from exc
        if len(raw) != size or hashlib.sha256(raw).hexdigest() != item.get('sha256'):
            raise ValueError('图片校验失败')
        out[name] = raw
    return out

File: core/lib/src/server/storage.dart
/// Validated, portable book snapshots and atomic JSON writes.
library;

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import '../errors.dart';
import '../pipeline/lang.dart';
import '../py/py_compat.dart';
import '../py/py_hash.dart';
import '../py/py_json.dart';
import '../py/py_re.dart';

const int assetLimit = 32 * 1024 * 1024;
const int assetsLimit = 96 * 1024 * 1024;
const int maxRecords = 1000000;
final RegExp _assetName = pyRe(r'[A-Za-z0-9_-][A-Za-z0-9_.-]{0,127}\Z');

/// `json.dump(value, ensure_ascii=False, separators=(',', ':'))` written
/// through a temporary file and an atomic rename.
void writeJson(File path, Object? value) {
  path.parent.createSync(recursive: true);
  final File tmp = File(
    '${path.parent.path}/.${path.uri.pathSegments.last}.$pid.${DateTime.now().microsecondsSinceEpoch}.tmp',
  );
  try {
    final RandomAccessFile out = tmp.openSync(mode: FileMode.write);
    try {
      out.writeFromSync(
        utf8.encode(PyJson.encode(value, ensureAscii: false, compact: true)),
      );
      out.flushSync();
    } finally {
      out.closeSync();
    }
    tmp.renameSync(path.path);
  } finally {
    if (tmp.existsSync()) tmp.deleteSync();
  }
}

final RegExp _byline = pyRe(
  r'^(?P<title>.*?)\s*[-_—\s]*作者\s*[:：]\s*(?P<author>.*?)\s*$',
);
final RegExp _site = pyRe(r'(?i)^[\w.-]+\.(com|net|org|cn|cc)$');

/// TXT file names often carry the byline ("书名 作者：某某"); show it as
/// title + author. Display only: stored book data stays unchanged.
(String, String) displayTitle(String? title, String? author) {
  String t = PyCompat.strip(title ?? '');
  String a = PyCompat.strip(author ?? '');
  final RegExpMatch? m = pyMatch(_byline, t);
  if (m != null && (m.namedGroup('title') ?? '').isNotEmpty) {
    t = m.namedGroup('title')!;
    final String found = PyCompat.strip(m.namedGroup('author') ?? '');
    if (a.isEmpty && found.isNotEmpty && pyMatch(_site, found) == null) {
      a = found;
    }
  }
  return (t, a);
}

bool _truthy(Object? v) =>
    v != null &&
    v != false &&
    v != 0 &&
    v != '' &&
    !(v is List<Object?> && v.isEmpty) &&
    !(v is Map<Object?, Object?> && v.isEmpty);

Map<String, Object?> shelfFields(Map<String, Object?> book) {
  final List<Object?> chapters =
      _truthy(book['chapters']) ? book['chapters']! as List<Object?> : const [];
  final Map<String, Object?> out = <String, Object?>{
    for (final String k in <String>[
      'title',
      'author',
      'len',
      'cover',
      'lang',
      'genre',
    ])
      k: book[k],
  };
  if (!_truthy(out['lang'])) out['lang'] = bookLang(book);
  final Object? len = book['len'];
  out['chapters'] = chapters.length;
  out['thin'] =
      (_truthy(len) ? len! as num : 0) < 30000 ||
      chapters
              .where(
                (Object? c) => (c! as Map<String, Object?>)['kind'] == 'body',
              )
              .length <
          2;
  return out;
}

int integer(Object? value, String label, {int low = 0, int high = 120000000}) {
  if (value is! int || value < low || value > high) {
    throw ValueError('$label无效');
  }
  return value;
}

String assetName(Object? name) {
  if (name is! String || pyFullmatch(_assetName, name) == null) {
    throw const ValueError('图片文件名无效');
  }
  return name;
}

int _utf16Len(String s) => s.length;

Map<String, Object?> validateBook(Object? raw) {
  if (raw is! Map<String, Object?> || raw['title'] is! String) {
    throw const ValueError('书籍信息无效');
  }
  final Map<String, Object?> book = raw;
  final int length = integer(book['len'], '正文长度');
  final Object? blocks = book['blocks'];
  final Object? chapters = book['chapters'];
  if (blocks is! List<Object?> ||
      blocks.isEmpty ||
      blocks.length > maxRecords) {
    throw const ValueError('正文段落无效');
  }
  if (chapters is! List<Object?> ||
      chapters.isEmpty ||
      chapters.length > 100000) {
    throw const ValueError('章节无效');
  }
  int previous = -1;
  for (final Object? block in blocks) {
    if (block is! Map<String, Object?> ||
        !<Object?>['p', 'h', 'img'].contains(block['k']) ||
        block['t'] is! String) {
      throw const ValueError('正文段落格式无效');
    }
    final String text = block['t']! as String;
    final int pos = integer(block['o'], '段落位置', high: length);
    if (pos < previous || pos + _utf16Len(text) > length) {
      throw const ValueError('段落位置超出正文');
    }
    previous = pos;
    if (block['k'] == 'img') assetName(block['src']);
    for (final Object? item in (block['fn'] as List<Object?>?) ?? const []) {
      if (item is! List<Object?> || item.length != 2 || item[1] is! String) {
        throw const ValueError('脚注索引无效');
      }
      integer(item[0], '脚注位置', high: _utf16Len(text));
    }
  }
  previous = -1;
  for (final Object? chapter in chapters) {
    if (chapter is! Map<String, Object?> || chapter['title'] is! String) {
      throw const ValueError('章节格式无效');
    }
    final int a = integer(chapter['b0'], '章节起点', high: blocks.length);
    final int b = integer(chapter['b1'], '章节终点', low: a, high: blocks.length);
    final int p = integer(chapter['o0'], '章节位置', high: length);
    integer(chapter['o1'], '章节终点', low: p, high: length);
    if (a == b || p < previous) throw const ValueError('章节顺序无效');
    previous = p;
  }
  final Object? notes =
      book.containsKey('notes') ? book['notes'] : <String, Object?>{};
  if (notes is! Map<String, Object?>) throw const ValueError('脚注格式无效');
  if (!notes.values.every((Object? v) => v is String)) {
    throw const ValueError('脚注内容无效');
  }
  if (_truthy(book['cover'])) assetName(book['cover']);
  return book;
}

const Map<String, List<String>> _graphFields = <String, List<String>>{
  'person': <String>['id', 'name'],
  'name': <String>['id', 'name'],
  'alias': <String>['id', 'alias'],
  'merge': <String>['from', 'into'],
  'profile': <String>['id'],
  'attr': <String>['id', 'key', 'value'],
  'rel': <String>['a', 'b'],
  'event': <String>['text'],
  'imp': <String>['id'],
  'cnt': <String>[],
  'recap': <String>['text'],
  'saga': <String>['text'],
};

Map<String, Object?> validateGraph(Object? raw, int length) {
  if (raw is! Map<String, Object?> ||
      (raw.containsKey('log') && raw['log'] is! List<Object?>)) {
    throw const ValueError('人物图谱格式无效');
  }
  final List<Object?> log = (raw['log'] as List<Object?>?) ?? const [];
  if (log.length > maxRecords) throw const ValueError('人物图谱太大');
  int previous = -1;
  for (final Object? item in log) {
    if (item is! Map<String, Object?> || !_graphFields.containsKey(item['t'])) {
      throw const ValueError('人物记录类型无效');
    }
    final Map<String, Object?> row = item;
    final int pos = integer(row['p'], '人物记录位置', high: length);
    if (pos < previous) throw const ValueError('人物记录没有按位置排序');
    previous = pos;
    for (final String key in _graphFields[row['t']]!) {
      if (row[key] is! String) throw const ValueError('人物记录字段无效');
    }
    for (final String key in const <String>[
      'intro',
      'tagline',
      'bio',
      'family',
      'a_is',
      'b_is',
      'desc',
      'status',
      'reason',
    ]) {
      if (row.containsKey(key) && row[key] != null && row[key] is! String) {
        throw const ValueError('人物记录文本无效');
      }
    }
    if (row.containsKey('s')) integer(row['s'], '原文位置', high: length);
    if (row.containsKey('imp')) integer(row['imp'], '人物重要度', high: 100);
    if (row.containsKey('chapter')) {
      integer(row['chapter'], '章节编号', high: 100000);
    }
    if (row['t'] == 'event') {
      final Object? who = row['who'];
      if (who is! List<Object?> || !who.every((Object? x) => x is String)) {
        throw const ValueError('事件人物无效');
      }
    }
    if (row['t'] == 'cnt') {
      final Object? c = row['c'];
      if (c is! Map<String, Object?> ||
          !c.values.every((Object? v) => v is int && v >= 0)) {
        throw const ValueError('人物计数无效');
      }
    }
  }
  return raw;
}

Set<String> referencedAssets(Map<String, Object?> book) {
  final Set<String> names = <String>{};
  for (final Object? raw in book['blocks']! as List<Object?>) {
    final Map<String, Object?> b = raw! as Map<String, Object?>;
    if (b['k'] == 'img') names.add(b['src']! as String);
  }
  if (_truthy(book['cover'])) names.add(book['cover']! as String);
  return names;
}

/// Base64 image payloads for a portable export, in sorted name order.
Map<String, Object?> encodeAssets(Directory root, Map<String, Object?> book) {
  final Map<String, Object?> assets = <String, Object?>{};
  int total = 0;
  final List<String> names = referencedAssets(book).toList()..sort();
  for (final String name in names) {
    assetName(name);
    final File path = File('${root.path}/img/$name');
    if (FileSystemEntity.isLinkSync(path.path) || !path.existsSync()) {
      throw const ValueError('这本书缺少图片，请先修复后再导出');
    }
    final int size = path.lengthSync();
    total += size;
    if (size > assetLimit || total > assetsLimit) {
      throw const ValueError('图片超过导出限制');
    }
    final Uint8List bytes = path.readAsBytesSync();
    assets[name] = <String, Object?>{
      'size': bytes.length,
      'sha256': sha256Hex(bytes),
      'base64': base64.encode(bytes),
    };
  }
  return assets;
}

final RegExp _base64 = RegExp(r'^[A-Za-z0-9+/]*={0,2}$');

Map<String, Uint8List> decodeAssets(Object? assets, Map<String, Object?> book) {
  if (assets is! Map<String, Object?> ||
      !_sameSet(assets.keys.toSet(), referencedAssets(book))) {
    throw const ValueError('图片清单与正文不一致');
  }
  final Map<String, Uint8List> out = <String, Uint8List>{};
  int total = 0;
  for (final MapEntry<String, Object?> e in assets.entries) {
    assetName(e.key);
    final Object? item = e.value;
    if (item is! Map<String, Object?>) throw const ValueError('图片格式无效');
    final int size = integer(item['size'], '图片长度', high: assetLimit);
    total += size;
    if (total > assetsLimit) throw const ValueError('图片总量太大');
    final Object? encoded = item['base64'];
    if (encoded is! String || encoded.length > ((size + 2) ~/ 3) * 4) {
      throw const ValueError('图片编码长度无效');
    }
    final Uint8List raw;
    try {
      if (encoded.length % 4 != 0 || !_base64.hasMatch(encoded)) {
        throw const FormatException();
      }
      raw = base64.decode(encoded);
    } on FormatException {
      throw const ValueError('图片编码无效');
    }
    if (raw.length != size || sha256Hex(raw) != item['sha256']) {
      throw const ValueError('图片校验失败');
    }
    out[e.key] = raw;
  }
  return out;
}

bool _sameSet(Set<String> a, Set<String> b) =>
    a.length == b.length && a.containsAll(b);
